File: src/main/java/com/vc/hard/L493.scala
package com.vc.hard

object L493 {
  def reversePairs(nums: Array[Int]): Int = {
    var count = 0
    def mergeSort(arr: Array[Int]): Array[Int] = {
      if(arr.length <= 1) return arr

      val lo = 0
      val hi = arr.length
      val mid = lo + (hi - lo) / 2
      val left = mergeSort(arr.slice(0, mid))
      val right = mergeSort(arr.slice(mid, hi))

      var leftCount = 0
      var rightCount = 0
      while(leftCount < left.length) {
        val leftElement = left(leftCount) / 2.0
        while(rightCount < right.length && leftElement > right(rightCount)) rightCount += 1
        count += rightCount
        leftCount += 1
      }

      leftCount = 0
      rightCount = 0
      while(leftCount < left.length && rightCount < right.length) {
        val leftElement = left(leftCount)
        val rightElement = right(rightCount)
        if(leftElement < rightElement) {
          arr(leftCount + rightCount) = leftElement
          leftCount += 1
        }
        else {
          arr(leftCount + rightCount) = rightElement
          rightCount += 1
        }
      }

      while(leftCount < left.length) {
        arr(leftCount + rightCount) = left(leftCount)
        leftCount += 1
      }

      while(rightCount < right.length) {
        arr(leftCount + rightCount) = right(rightCount)
        rightCount += 1
      }
      arr
    }
    mergeSort(nums)
    count
  }
}

File: src/main/java/com/vc/hard/L493.java
package com.vc.hard;

class L493 {
    int count = 0;
    public int reversePairs(int[] nums) {
        mergeSort(nums);
        return count;
    }

    private int[] mergeSort(int[] arr) {
        if(arr.length <= 1) return arr;

        int lo = 0;
        int hi = arr.length;
        int mid = lo + (hi - lo) / 2;
        int[] left = mergeSort(slice(arr, lo, mid));
        int[] right = mergeSort(slice(arr, mid, hi));

        int leftCount = 0;
        int rightCount = 0;
        while(leftCount < left.length) {
            double leftElement = left[leftCount] / 2.0;
            while(rightCount < right.length && leftElement > right[rightCount]) rightCount++;
            leftCount++;
            count += rightCount;
        }

        leftCount = 0;
        rightCount = 0;
        while(leftCount < left.length && rightCount < right.length) {
            if(left[leftCount] < right[rightCount])
                arr[leftCount + rightCount] = left[leftCount++];
            else
                arr[leftCount + rightCount] = right[rightCount++];
        }

        while(leftCount < left.length) {
            arr[leftCount + rightCount] = left[leftCount++];
        }

        while(rightCount < right.length) {
            arr[leftCount + rightCount] = right[rightCount++];
        }

        return arr;
    }

    private int[] slice(int[] arr, int from, int to) {
        if(from >= arr.length) return new int[0];
        int[] res = new int[to - from];
        int index = 0;
        for(int i = from; i < to; i++) {
            res[index++] = arr[i];
        }
        return res;
    }
}
